[pytorch] Avoid sending large unneeded data over wire in process_group_agent.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31357 [pytorch] Avoid sending large unneeded data over wire in process_group_agent.**

If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

Commit:
[pytorch] Avoid sending large unneeded data over wire in process_group_agent.

If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Avoid sending large unneeded data over wire in process_group_agent."


If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Avoid sending large unneeded data over wire in process_group_agent."


If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Avoid sending large unneeded data over wire in process_group_agent."


If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Avoid sending large unneeded data over wire in process_group_agent."


If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]

pritamdamania87 reviewed (changes requested):
Can we consolidate our wire serialization code across multiple backends? Adding similar functionality for both agents in separate places is going to be hard to maintain going forward. Can we refactor everything to ensure we reuse the same logic as much as possible?

jjlilley commented:
> Can we consolidate our wire serialization code across multiple backends? Adding similar functionality for both agents in separate places is going to be hard to maintain going forward. Can we refactor everything to ensure we reuse the same logic as much as possible?

Ideally we would.

However, AFAICT, we have the following constraints:
 - process_group_agent, per policy reasons, isn't allowed to depend on folly (if that changes, awesome)
- Thrift, OTOH, primarily uses folly::IOBuf for buffering, and for avoiding some unnecessary buffer copies.
- Do we care about avoiding some unnecessary ThriftRpcAgent memory copies? I suspect that we do.

If those constraints are all correct, I don't see a viable alternative to producing separate serializers that work on IOBufs/non-IOBufs. Especially when it's really largely glue code around jit::Pickler - i.e. the OSS wireSerialize is ~70 lines of code. If there's a realistic alternative, let me know.

Keep in mind, too, that the num serializers here isn't O(num_backends), but rather O(num_backend_memory_formats). If we allowed folly in OSS, this would be back to one.

satgera commented on test/cpp/rpc/test_wire_serialization.cpp:
```diff
@@ -37,3 +37,16 @@ TEST(WireSerialize, Base) {
+
+TEST(WireSerialize, RecopySparseTensors) {
+  // Take a 1K row of a 1M tensors, and make sure we don't send across 1M rows.
+  static const size_t k1K = 1024;
```
Nit: Prefer using constexpr instead of static const.

jjlilley replied:
Just replaced a few constexpr
Sorry, 'static const' is a bit of an ingrained habit, like C-x C-s/etc :)

satgera commented:
> Can we consolidate our wire serialization code across multiple backends? Adding similar functionality for both agents in separate places is going to be hard to maintain going forward. Can we refactor everything to ensure we reuse the same logic as much as possible?

Pritam, I agree in principle we should keep least amount of parallel logic, should we open an issue to see what ideas we can come up with and let this diff go in ?

Commit:
Update on "[pytorch] Avoid sending large unneeded data over wire in process_group_agent."


If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]

jjlilley commented:
btw, re: convergence with OSS and fb-only version, I did just amend a diff for the FB-only version to change its vector<Tensor> arg to const ref from move semantics (since in practice, we weren't able to take advantage of this).

Ideally, the two versions will converge sufficiently to have parts worth reusing, but my feeling is that we're not quite there yet.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 7a6274e:
* **1/1** broken upstream at merge base 3e59e80 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=3e59e8042970bbf8b5e207d9f26663a8cb1a8248&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7a6274e9c781f463b085d50062040d5f837f7288).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_xla_linux_xenial_py3_6_clang7_build](https://dr.pytorch.org/build-details.html?build_id=76756735)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pritamdamania87 commented:
I understand we can't converge the two codepaths easily, but can we at least refactor this particular code block into a common place that both code paths use? As far as I understand right now this is pretty much a copy of what we have in the ThriftAgent.

jjlilley commented:
> I understand we can't converge the two codepaths easily, but can we at least refactor this particular code block into a common place that both code paths use? As far as I understand right now this is pretty much a copy of what we have in the ThriftAgent.

Pritam, I'll come find you - it might be more constructive to chat offline.

pritamdamania87 reviewed (approved):
Spoke with @jjlilley offline, we can refactor and consolidate the `worthRecopying` part in a separate PR. But lets make sure we do this soon and it doesn't get dropped amongst other things :) 

Created an issue for the same: https://github.com/pytorch/pytorch/issues/31447

jjlilley commented:
> Spoke with @jjlilley offline, we can refactor and consolidate the `worthRecopying` part in a separate PR. But lets make sure we do this soon and it doesn't get dropped amongst other things :)
> 
> Created an issue for the same: #31447

Thanks. So, from chatting offline:
 - it sounded like the main part that Pritam was concerned about was largely the worthRecopying portion.
-  I was concerned about trying to consolidate rest of the function, which right now, follows a similar form, but I think diverges significantly enough that it would be hard to merge.

If that's the case, I think it might make sense for me to make a quick amendment to this PR to break those ~15 lines of code into a separate exported cloneSparseTensors() function (no actual code change, just 2 funcs). Then the followup on the Thrift side to use the logic.

Commit:
Update on "[pytorch] Avoid sending large unneeded data over wire in process_group_agent."


If a user selects a subset of a Tensor and sends it in an RPC, we were sending
the whole original Tensor Storage over the network.

While this sounds reasonable, in practice, we observed view-like Tensors being sent
over rpc, where only 1% of the data in the provided Tensor's Storage was
actually used/needed.

The simple solution here is to just force a clone in the serializer code if we see that
less than (arbitrary) half the bits are used, and the tensor is more than a nominal few KB.
Add related tests to ensure this doesn't break.

An alternate approach would be to modify the Pickler. That said, since Pickler is shared by more
components, the logic might be harder to tailor appropriately at that layer (particularly
given that the Pickler has explicit logic to share a single Storage* among several Tensors
that commonly point to the same Storage*).

It's possible that we might want to further refine the basic thresholds in this change.
In practice, we've seen a mostly bimodal distribution thus far for the percent of Tensor
Storage referred by a Tensor in observed rpcs (i.e. either 90%+ or sub-10% of the Storage
referenced), hence the existing 50% threshold here is probably not an unreasonable
starting point.

Differential Revision: [D19137056](https://our.internmc.facebook.com/intern/diff/D19137056/)

[ghstack-poisoned]
